Mepts 79 updated breastfeeding and pregnant queries

ningosi opened the pull request:
@nthfloor here the PR for review

edrisse commented on api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/pvls/BreastfeedingAndPregnantCalculation.java:
```diff
@@ -0,0 +1,267 @@
+				// get patient_state for pregnant and breastfeeding
+				if (patientProgram != null) {
+					for (PatientState patientState : patientProgram.getCurrentStates()) {
+						if (patientState.getState().getProgramWorkflowStateId() == 27) {
```
This `program_workflow_state_ids` should be defined as constants on the global properties so that if a specific partner uses a different ID they can just configure it. That way we can support all partners with same codebase.

ningosi replied:
Done

edrisse commented on api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/pvls/BreastfeedingAndPregnantCalculation.java:
```diff
@@ -0,0 +1,267 @@
+					if (markedPregnantList.size() > 0) {
+						for (Obs obs : markedPregnantList) {
+							if (obs != null && obs.getObsDatetime() != null
+							        && (obs.getObsDatetime().equals(pregnancyStartDate)
```
I suggest we use `obs.getObsDatetime().compareTo(pregnancyStartDate) >= 0`. That way the if statement will have only one expression and be a bit more readable.

ningosi replied:
Done

Guimino commented on api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/pvls/BreastfeedingAndPregnantCalculation.java:
```diff
@@ -0,0 +1,252 @@
+				}
+				
+				if (criteria.equals(BreastfeedingAndPregnant.PREGNANT)) {
+					if (markedPregnantList.size() > 0) {
```
Use !isEmpty() to check whether the collection is empty or not.

Guimino commented on api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/pvls/BreastfeedingAndPregnantCalculation.java:
```diff
@@ -0,0 +1,252 @@
+					        && (breastfeedingState.getStartDate().compareTo(endDate) <= 0)) {
+						pass = true;
+					}
+					if (deliveryDateMapList.size() > 0) {
```
Use !isEmpty()  instead

Guimino commented on api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/pvls/BreastfeedingAndPregnantCalculation.java:
```diff
@@ -0,0 +1,252 @@
+						}
+					}
+					
+					if (markedPregnantOnEddList.size() > 0) {
```
Use !isEmpty()  instead

Guimino commented on api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/pvls/BreastfeedingAndPregnantCalculation.java:
```diff
@@ -0,0 +1,252 @@
+							}
+						}
+					}
+					if (lactatingMapList.size() > 0) {
```
Use !isEmpty()  instead

Guimino commented on api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/pvls/BreastfeedingAndPregnantCalculation.java:
```diff
@@ -0,0 +1,252 @@
+							}
+						}
+					}
+					if (criteriaHivStartMapList.size() > 0) {
```
Use !isEmpty() instead

Guimino commented on api/src/main/java/org/openmrs/module/eptsreports/metadata/HivMetadata.java:
```diff
@@ -162,5 +162,17 @@ public ProgramWorkflowState getPatientHasDiedWorkflowState() {
 		    EptsReportConstants.GLOBAL_PROPERTY_ART_PROGRAM_UUID);
 		return getProgramWorkflowState(artProgramUuid, "2", "PATIENT HAS DIED");
 	}
+
```
execute mvn clean install before committing the changes in order to have formatted code

Guimino commented on api/src/main/java/org/openmrs/module/eptsreports/reporting/library/datasets/TxPvlsDataset.java:
```diff
@@ -61,6 +62,54 @@ public DataSetDefinition constructTxPvlsDatset() {
 		        eptsGeneralIndicator.getIndicator("suppressed viral load",
 		            EptsReportUtils.map(txPvls.getPatientsWithViralLoadSuppression(), mappings)), mappings), "");
 		
+		// Columns useful for debugging totals
```
please remove this block of commented lines of code

ningosi commented:
@Guimino addressed your concerns, but seems I need to find a away of splitting this long method into smaller chunks for readability and quality improvement.

ynurmahomed commented on api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxPvlsCohortQueries.java:
```diff
@@ -121,11 +123,12 @@ public CohortDefinition getPatientsWithViralLoadSuppression() {
         EptsReportUtils.map(
             getPatientsWhoAreMoreThan3MonthsOnArt(), "onDate=${endDate},location=${location}"));
-    cd.setCompositionString("supp AND onArtLongEnough");
+    cd.setCompositionString("supp AND baseCohort AND onArtLongEnough");
```
Isn't `baseCohort` defined at `ReportDefinition` level,  does it need to be defined here also?

Guimino replied:
removed baseCohort from the remain composition String

ynurmahomed commented on api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxPvlsCohortQueries.java:
```diff
@@ -346,11 +350,18 @@ public CohortDefinition getPregnantWomenWithViralLoadSuppressionNumerator() {
     cd.addSearch(
         "suppression", EptsReportUtils.map(getPatientsWithViralLoadSuppression(), mappings));
+
+    // cd.addSearch("pregnant", EptsReportUtils.map(getPregnantSQLQuery(),
```
Please delete commented code.

Guimino replied:
removed

ynurmahomed commented on api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/pvls/PregnantCalculation.java:
```diff
@@ -0,0 +1,217 @@
+    for (Integer pId : femaleCohort) {
+
+      Boolean isCandidate = false;
+      if (onArtForMoreThan3Months.contains(pId)) {
```
I wonder why the patient needs to be on ART for more than 3 months to be considered pregnant. 
This calculation will only be valid for `TX_PVLS` right? 
It can stay the way it is, but would it be better to have this condition on the composition cohort?

Guimino replied:
@ynurmahomed, **good question**.I realized that the verification of  on ART for more than 3 months is made in others cohorts (other parts of the code)  and when needed the result is crossed with the pregnant/breastfeeding result, I removed this verification and the result remain the same, I will remove this verification

ynurmahomed replied:
:+1:

ynurmahomed commented on api/src/main/java/org/openmrs/module/eptsreports/reporting/utils/EptsReportConstants.java:
```diff
@@ -139,4 +139,9 @@
     ADULTCHILDREN
   }
+
+  public enum BreastfeedingAndPregnant {
```
Looks like there are no usages for this enum, please delete.

Guimino replied:
@ynurmahomed , done

ynurmahomed commented on api/src/main/java/org/openmrs/module/eptsreports/reporting/reports/SetupTxPvls.java:
```diff
@@ -28,7 +26,6 @@
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Component;
 
-@Deprecated
```
I think `@Deprecated` should stand.

Guimino replied:
Done

Guimino commented:
@ynurmahomed , about the two issues you mentioned that are not related to the changes:

I Addressed the second (`base` cohort is added twice to the `TX_PVLS` report definition) tests were performed to make sure the results remain the same.

nthfloor commented on api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxPvlsCohortQueries.java:
```diff
@@ -527,4 +527,50 @@ public CohortDefinition getPregnantWomenAndNotDocumentedWithViralLoadResultsDeno
+    return cd;
+  }
+
+  public CohortDefinition getPatientsWhoArePregnant() {
```
Can we please rename this function to be more descriptive of what it is doing because right now this is too similar to the function below: `getPatientsWhoArePregnantCohort`

Maybe something like `getPatientsWhoArePregnantAndNotBreastfeeding`

Guimino replied:
ok @nthfloor , will rename to the name you suggested

Guimino replied:
done

nthfloor commented on api/src/main/java/org/openmrs/module/eptsreports/reporting/library/datasets/TxPvlsDataset.java:
```diff
@@ -63,6 +61,69 @@ public DataSetDefinition constructTxPvlsDatset() {
             mappings),
         "");
 
+    dsd.addColumn(
```
What is the reason for adding these extra columns? From what I remember this change was only to change the period...

Guimino replied:
@nthfloor ,I think some one added these two columns for test purpose and forgot to delete 
(

> dsd.addColumn("P", "Pregnant". , dsd.addColumn("B", "Breastfeeding")

Guimino replied:
I will remove these columns

nthfloor replied:
@Guimino these columns are still in this dataset?

Guimino replied:
@nthfloor  I removed them from the dataset

nthfloor replied:
really I still see them when I pull latest...

Guimino replied:
@nthfloor , make sure you are pulling the latest, @ningosi  can you see that columns ?

nthfloor commented:
@Guimino please also look into adding tests for these two new calculations. You can pull in the changes from METPS-83 which will be merged soon. Kaweesi has added some tests for the routine, initiateART and OnARTFOr3months calculations already.

Guimino commented:
ok @nthfloor , will add unit tests for the calculations

ningosi commented:
Also could the patient state start date be forced to be the same as program enrollment date?

ningosi commented on api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/pvls/PregnantCalculation.java:
```diff
@@ -171,7 +170,6 @@ private boolean isPregnantDueDate(Date lastVlDate, List<Obs> pregnantDueDateObsL
-
     return patientState != null
         && patientState.getStartDate() != null
         && patientState.getEndDate() == null
```
What if the patientState.getEndDate() has value that falls within the reporting period or the dates ranges specified? will that be counted or we completely need this column to be null?

Guimino replied:
@ningosi , now I'm considering patient Program.dateEnrolled to compare with the period range, I think using patientState.startDate could be the mistake in the pregnantCalculator

> Patients that are female and enrolled on PTV/ETC program during the period range.

ningosi commented:
@Guimino your changes/work looks great! please verify for me that the number you get by getting patients who are enrolled in PTV program with the states pregnancy and breastfeeding is the same as those gotten by writing native sql for that matter, the rest of the calls are fine.

Guimino commented:
@ningosi , I'm looking at this

kaweesi commented:
please resolve the conflicts above.

kaweesi commented on api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/AbstractPatientCalculation.java:
```diff
@@ -57,4 +59,12 @@ protected static CalculationResultMap calculate(
   }
+
+  protected HivMetadata getHivMetadata() {
+    return Context.getRegisteredComponents(HivMetadata.class).get(0);
```
@nthfloor rejected this aTM, we reverted it from the unit test framework. please go ahead and update to match master

Guimino replied:
this is no longer there after merge

kaweesi commented on api/src/main/java/org/openmrs/module/eptsreports/reporting/cohort/definition/CalculationCohortDefinition.java:
```diff
@@ -1,17 +1,17 @@
+ * License at http://license.openmrs.org ======= /* The contents of this file are subject to the
+ * OpenMRS Public License Version 1.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at http://license.openmrs.org
+ * >>>>>>> master
```
this looks like merge conflict marks

nthfloor replied:
@Guimino this still needs to be addressed

Guimino replied:
fixed

kaweesi commented on api/src/test/java/org/openmrs/module/eptsreports/reporting/helper/OpenMRSTestHelper.java:
```diff
@@ -0,0 +1,38 @@
+import org.openmrs.Patient;
+import org.openmrs.api.context.Context;
+
+public class OpenMRSTestHelper {
```
this is already in master

kaweesi commented on api/src/test/java/org/openmrs/module/eptsreports/reporting/helper/TestsHelper.java:
```diff
@@ -0,0 +1,34 @@
+package org.openmrs.module.eptsreports.reporting.helper;
+
```
this too

kaweesi commented on api/src/test/java/org/openmrs/module/eptsreports/reporting/intergrated/calculation/BasePatientCalculationTest.java:
```diff
@@ -0,0 +1,135 @@
+import org.openmrs.calculation.patient.PatientCalculationService;
+import org.openmrs.calculation.result.CalculationResult;
+import org.openmrs.calculation.result.CalculationResultMap;
+import org.openmrs.module.eptsreports.api.EptsReportsService;
```
this as well, probably u just didn't branch off the test framework branch, try doing that so save u the pain of having to reset and throw away all these

Guimino commented on api/pom.xml:
```diff
@@ -21,10 +21,30 @@
 		</dependency>
+		<dependency>
+        	<groupId>org.openmrs.module</groupId>
+            <artifactId>serialization.xstream-api-2.0</artifactId>
```
conflicts resolved, I'll remove this entry in pom.xml

Guimino replied:
this is no longer there after merge

Guimino commented on api/pom.xml:
```diff
@@ -21,10 +21,30 @@
 		</dependency>
+		<dependency>
+			<groupId>org.openmrs.module</groupId>
+			<artifactId>calculation-api</artifactId>
```
no longer there after merge

Guimino commented on api/pom.xml:
```diff
@@ -21,10 +21,30 @@
+		
+		<dependency>
+		    <groupId>org.jmockit</groupId>
+		    <artifactId>jmockit</artifactId>
```
this is no longer there after merge

nthfloor commented on api/src/main/java/org/openmrs/module/eptsreports/reporting/utils/EptsCalculationUtils.java:
```diff
@@ -285,4 +285,40 @@ public static Date addMonths(Date endDate, int months) {
+   * @param context the calculation context
+   * @return the genders in a calculation result map
+   */
+  public static CalculationResultMap genders(
```
Do we have unit tests for these functions?

kaweesi commented on api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/EptsCalculations.java:
```diff
@@ -35,7 +34,10 @@
-/** Utility class of common base calculations */
+/**
+ * Utility class of common base calculations TODO: refactor needs to be merged with
+ * EptsCalculationUtils
```
so why not merge them rather, do we have such a ticket @nthfloor ?

nthfloor replied:
When I added that comment I was thinking we should merge. But seeing now how those two classes have evolved I don't think we should merge anymore. 
* EptsCalculationUtils is currently a collections of helper functions 
* Whereas EptsCalculations has lookup functions for getting Obs and encounters. Kind of like a service for looking up openmrs objects (we should probably rename to be more descriptive)

kaweesi commented on api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/EptsCalculations.java:
```diff
@@ -152,25 +154,42 @@ public static CalculationResultMap lastObs(
+
+    JembiPatientStateDefinition def = new JembiPatientStateDefinition();
+    def.setLocation(location);
+    def.setStartedOnOrBefore(context.getNow());
```
patientState startdate being now!!! i think i need to re-look through the requirements.

ningosi commented:
Context.getNow() is later passed to reporting end date period is supplied
otherwise today's date is passed. Look at the CalculationCohortDefinion
class that evaluates calculations

On Fri, Feb 22, 2019, 18:22 Kaweesi Joseph <notifications@github.com wrote:

> *@kaweesi* commented on this pull request.
> ------------------------------
>
> In
> api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/EptsCalculations.java
> <https://github.com/esaude/openmrs-module-eptsreports/pull/40#discussion_r259385704>
> :
>
> >     */
> -  public static CalculationResultMap firstProgramEnrollment(
> +  public static CalculationResultMap allPatientStates(
> +      Collection<Integer> cohort,
> +      Location location,
> +      ProgramWorkflowState programWorkflowState,
> +      PatientCalculationContext context) {
> +
> +    JembiPatientStateDefinition def = new JembiPatientStateDefinition();
> +    def.setLocation(location);
> +    def.setStartedOnOrBefore(context.getNow());
>
> patientState startdate being now!!! i think i need to re-look through the
> requirements.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/esaude/openmrs-module-eptsreports/pull/40#pullrequestreview-206873855>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABrn7AJ5YR2qbh_EPQtzqsCH_PaAUgMUks5vQAtPgaJpZM4aScye>
> .
>

nthfloor commented on api/src/test/java/org/openmrs/module/eptsreports/reporting/unit/calculation/EptsCalculationsTest.java:
```diff
@@ -1,3 +1,42 @@
+import org.openmrs.module.eptsreports.reporting.utils.EptsCalculationUtils;
+import org.openmrs.test.BaseModuleContextSensitiveTest;
+
+public class EptsCalculationsTest extends BaseModuleContextSensitiveTest {
```
@ningosi this is technically an integration test since it spins up a context. 

The reason I think we should keep these separate from unit tests is that they take longer to execute and we might want to eventually setup a separate test goal for testing these separately to our unit tests.

nthfloor replied:
Also I see this is infact for EptsCalculationsUtils so test class should be renamed to: EptsCalculationsTest.

Please rename and move to the integration test folder. Thanks :+1:

nthfloor replied:
Please also resolve codacy warning
